Detectors/TPC/reconstruction/src/HwClusterer.cxx

```diff
@@ -8,461 +8,530 @@
-          }
-
-          /*
-           * search for clusters and store reference to CF if one was found
-           */
-          if (clusterFinder[iCRU][iRow][0]->getTimebinsAfterLastProcessing() == iTimebinsPerCF-2 -2)  {
-            /*
-             * ordering is important: from right to left, so that the CFs could inform each other if cluster was found
-             */
-            for (auto rit = clusterFinder[iCRU][iRow].crbegin(); rit != clusterFinder[iCRU][iRow].crend(); ++rit) {
-              if ((*rit)->findCluster()) {
-                cfWithCluster.push_back(rit);
-              }
-            }
-          }
-        }
-        for (auto rit = clusterFinder[iCRU][iRow].crbegin(); rit != clusterFinder[iCRU][iRow].crend(); ++rit) {
-          (*rit)->setTimebinsAfterLastProcessing(0);
+        HB = i < 2 ? 0 : (i - 2) / 447; // integer division on purpose
+        if (HB != mLastHB) {
+          writeOutputWithTimeOffset(mLastHB * 447);
         }
-      }
 
-      /*
-       * collect found cluster
-       */
-      for (auto &cf_rit : cfWithCluster) {
-        auto cc = (*cf_rit)->getClusterContainer();
-        for (auto& c : *cc) cluster[iCRU].push_back(c);
+        /*
+         * For each row, we first check for cluster peaks in the timebin i-2,
+         * or because (i-2) mod 5 = (i+3) mod 5, in i+3, to ensure positive times.
+         *
+         * If timebin 5 is the new digit timebin (4 would be mLastTimebin),
+         * then 0 is the oldest one timebin and will to be replaced by the new
+         * arriving one. The cluster which could be found, would then range
+         * from timebin 0 to 4 and hasits center at timebin 2. Threrefore we
+         * are looking in (i - 2) for clusters and clearing (i - 4), or (i + 1)
+         * afterwards.
+         *       ---------
+         * -> 0 |
+         *    1 |
+         *    2 | XXXXXX
+         *    3 |
+         *    4 |
+         *       ---------
+         */
+        findClusterForTime(i - 2);
 
-        auto ll = (*cf_rit)->getClusterDigitIndices();
-        for (auto& l : *ll) {
-          label[iCRU].push_back(l);
-        }
+        clearBuffer(i + 1);
 
-        (*cf_rit)->clearClusterContainer();
+        mLastHB = HB;
       }
 
+      // we have to copy the MC truth container because we need the information
+      // maybe only in the next events (we store permanently 5 timebins), where
+      // the original pointer could already point to the next container.
+      if (mcDigitTruth) {
+        if (mCurrentMcContainerInBuffer == 0)
+          mMCtruth[digit.getTimeStamp() % 5] = std::make_shared<MCLabelContainer const>(*mcDigitTruth);
+        else
+          mMCtruth[digit.getTimeStamp() % 5] = std::shared_ptr<MCLabelContainer const>(mMCtruth[getFirstSetBitOfField()]);
+
+        mCurrentMcContainerInBuffer |= (0x1 << (digit.getTimeStamp() % 5));
+      } else
+        mMCtruth[digit.getTimeStamp() % 5] = std::shared_ptr<MCLabelContainer const>(nullptr);
```

Comment:
Is this last else needed at all? since mMCtruth is a `vector` of `shared_ptr` it should be initialized to `nullptr` by default I assume. If it is needed, please put the statement in `{}` even if it is a one-liner. This should always be done for readability and is also in the coding conventions I think.